File: app/src/main/java/org/skyfaced/mvp/ui/base/BaseActivity.java
package org.skyfaced.mvp.ui.base;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.viewbinding.ViewBinding;

import com.google.android.material.snackbar.Snackbar;

import moxy.MvpAppCompatActivity;

public abstract class BaseActivity<VB extends ViewBinding> extends MvpAppCompatActivity implements BaseView {
    private VB binding = null;

    @NonNull
    protected VB getBinding() {
        if (binding == null)
            throw new IllegalArgumentException("You are trying to access the binding after onDestroy()?");
        return binding;
    }

    @NonNull
    protected abstract VB setupBinding(LayoutInflater inflater);

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = setupBinding(getLayoutInflater());
        setContentView(binding.getRoot());
    }

    @Override
    protected void onDestroy() {
        binding = null;
        super.onDestroy();
    }

    /************
     * ViewImpl *
     ************/
    @Override
    public void hideKeyboard() {
        InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
    }

    @Override
    public void showToast(@NonNull String message, Integer duration) {
        Toast.makeText(this, message, duration).show();
    }

    @Override
    public void showToast(@NonNull Integer message, Integer duration) {
        Toast.makeText(this, message, duration).show();
    }

    @Override
    public void showSnackbar(@NonNull String message, Integer duration) {
        Snackbar.make(getBinding().getRoot(), message, duration).show();
    }

    @Override
    public void showSnackbar(@NonNull Integer message, Integer duration) {
        Snackbar.make(getBinding().getRoot(), message, duration).show();
    }
}

File: app/src/main/java/org/skyfaced/mvp/ui/base/BaseActivity.kt
package org.skyfaced.mvp.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.viewbinding.ViewBinding
import com.google.android.material.snackbar.Snackbar
import moxy.MvpAppCompatActivity

abstract class BaseActivity<VB : ViewBinding> : MvpAppCompatActivity(), BaseView {
    private var _binding: VB? = null
    protected val binding: VB
        get() = _binding
            ?: throw IllegalStateException("You are trying to access the binding after onDestroy()?")

    protected abstract fun setupBinding(inflater: LayoutInflater): VB

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = setupBinding(layoutInflater)
        setContentView(binding.root)
    }

    override fun onDestroy() {
        _binding = null
        super.onDestroy()
    }

    override fun hideKeyboard() {
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(currentFocus?.windowToken, 0)
    }

    override fun showToast(message: String, duration: Int) {
        Toast.makeText(this, message, duration).show()
    }

    override fun showToast(message: Int, duration: Int) {
        Toast.makeText(this, message, duration).show()
    }

    override fun showSnackbar(message: String, duration: Int) {
        Snackbar.make(binding.root, message, duration).show()
    }

    override fun showSnackbar(message: Int, duration: Int) {
        Snackbar.make(binding.root, message, duration).show()
    }
}